juju/juju
# Charm Platform Origin and Bundles

The following updates the deploy code to allow for platform to reside in
origin. With this change, it's now possible to correctly wire up the bundle
support and model migration changes. This helps with the wire up of the
architecture to platform so we can effectively run model migrations with
architecture support.

### Architecture

Architecture is wired through the deployment code by using the arch
constraint supplied and pushing it into the charm origin and pushing
that to the server.

There is a hack where the arch sent to the charmhub is hardcoded to
"all" because of a major issue in the API that is currently being resolved.

### Bundle support

Bundle support is wired up so that we call the FindDownloadURL on the
server. The bundle URL along with the origin is passed back so we can
extract it from the client. Originally I was going to create a generic repo
adapter, but it causes even more churn. Instead, I create one locally that
correctly loads the bundle and returns it to the deployer.

### Selector

The selector code that was crafted for downloads was repurposed for
finding the correct bundle URL. It was moved to `charmhub/selector` so
that client and server could use it.

There is a gap in the selector code that will be fixed in another PR that
will fix the issue around revisions and selecting the correct one based on
either the charm URL or the origin.

### Model migration

The changes here just exposed the platform to the model migration.

## QA steps

There are no bundles that have charmhub charms in them. This PR is only
to ensure we get the correct bundle.

```sh
$ export JUJU_DEV_FEATURE_FLAGS=charm-hub
$ juju bootstrap lxd test --config charm-hub-url="https://api.staging.snapcraft.io"
$ juju deploy wiki-simple
Located bundle "ch:wiki-simple-4"
Resolving charm via charmstore: cs:trusty/mysql-55
Resolving charm via charmstore: cs:trusty/mediawiki-5
Executing changes:
- upload charm cs:mysql-55 for series trusty
- deploy application mysql on trusty using cs:mysql-55
- set annotations for mysql
- upload charm cs:trusty/mediawiki-5 for series trusty
- deploy application wiki on trusty using cs:trusty/mediawiki-5
- set annotations for wiki
- add relation wiki:db - mysql:db
- add unit mysql/0 to new machine 0
- add unit wiki/0 to new machine 1
Deploy of bundle completed.

```

### Fallback architecture and series.

```sh
$ export JUJU_DEV_FEATURE_FLAGS=charm-hub
$ juju bootstrap lxd test
$ juju deploy ubuntu
```

### TBA

## Documentation changes

Architecture is now funneled through the cli and to the controller so we download
the correct charm binary blob.
